flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -565,20 +563,37 @@ private void startTriggeringCheckpoint(CheckpointTriggerRequest request) {
                                                     request.isPeriodic,
-                                                    checkpointInfo.f1.checkpointId,
-                                                    checkpointInfo.f1.checkpointStorageLocation,
+                                                    checkpointInfo.f1,
                                                     request.getOnCompletionFuture()),
                                     timer);
 
             final CompletableFuture<?> coordinatorCheckpointsComplete =
-                    pendingCheckpointCompletableFuture.thenComposeAsync(
-                            (pendingCheckpoint) ->
-                                    OperatorCoordinatorCheckpoints
-                                            .triggerAndAcknowledgeAllCoordinatorCheckpointsWithCompletion(
-                                                    coordinatorsToCheckpoint,
-                                                    pendingCheckpoint,
-                                                    timer),
-                            timer);
+                    pendingCheckpointCompletableFuture
+                            .thenApplyAsync(
+                                    pendingCheckpoint -> {
+                                        try {
+                                            CheckpointStorageLocation checkpointStorageLocation =
+                                                    initializeCheckpointLocation(
+                                                            pendingCheckpoint.getCheckpointID(),
+                                                            request.props,
+                                                            request.externalSavepointLocation,
+                                                            initializeBaseLocations);
+                                            pendingCheckpoint.setCheckpointTargetLocation(
+                                                    checkpointStorageLocation);
+                                            return pendingCheckpoint;
+                                        } catch (Throwable e) {
+                                            throw new CompletionException(e);
+                                        }
+                                    },
+                                    executor)
+                            .thenComposeAsync(
+                                    (pendingCheckpoint) ->
+                                            OperatorCoordinatorCheckpoints
+                                                    .triggerAndAcknowledgeAllCoordinatorCheckpointsWithCompletion(
+                                                            coordinatorsToCheckpoint,
+                                                            pendingCheckpoint,
+                                                            timer),
+                                    timer);
```

Comment:
I think that it is better to extract changes of moving the checkpoint storage location to a separate commit(the same PR/branch but the second commit). I think it will be better since the logging/statistic fixes and these changes resolve the different problems. But it is better to fix only one problem in one commit.